### PR TITLE
Introduce `Repository.MergeCommits`

### PR DESCRIPTION
**Pull request by ethomson**
This introduces `Repository.MergeCommits` which wraps `git_merge_commits` and is suitable for taking two commits and getting the merge results as an `Index`.  There's a handful of interesting details here:

* The conflict data previously had a pointer to the `Repository`.  This meant that they always went to the repository's `Index`.  I broke this and pass in the `Index` that they're supposed to pay attention to.  This is probably the least contentious change.

* Since the resultant `index` needs to be freed with `git_index_free`, I made `Index` implement `IDisposable` so that consumers of `Repository.MergeCommits` can just wrap that in a `using`.  To prevent anybody from trying to `Dispose` the `Repository.Index`, I only wired up `Dispose` for things that came from a merge.  

* I refactored `MergeOptions`, `CherryPickOptions` and `RevertOptions` to share a common base class.  They were getting a little copy-pastey.  I also pulled out the `git_merge_options`-like things into their own base class so that `MergeTreeOptions` could also take advantage.

I suspect there's going to be something in here for everybody to hate!  :)  Other suggestions for how to implement these things  are welcome.

### CODE REVIEW

File: LibGit2Sharp/Repository.cs [hunk 1]
```diff
@@ -1077,6 +1077,42 @@ public MergeResult MergeFetchedRefs(Signature merger, MergeOptions options)
...
+        /// Perform a three-way merge of two commits, looking up their
+        /// commit ancestor. The returned index will contain the results
+        /// of the merge and can be examined for conflicts. The returned
+        /// index must be disposed.
+        /// </summary>
+        /// <param name="ours">The first tree</param>
+        /// <param name="theirs">The second tree</param>
+        /// <param name="options">The <see cref="MergeTreeOptions"/> controlling the merge</param>
+        /// <returns>The <see cref="Index"/> containing the merged trees and any conflicts</returns>
+        public Index MergeCommits(Commit ours, Commit theirs, MergeTreeOptions options)
```
Comment (nulltoken) [inline LibGit2Sharp/Repository.cs]: Couldn't we move this to `repo.ObjectDatabase`?

Comment (ethomson) [inline LibGit2Sharp/Repository.cs]: Sure. Do you like `ObjectDatabase.MergeTrees`?

Comment (nulltoken) [inline LibGit2Sharp/Repository.cs]: :+1:

Comment (nulltoken) [inline LibGit2Sharp/Repository.cs]: However, I'm not sure about the returned parameter type. I wonder if we should return a different type (an interface?). 

How would *this** index behave when being added/removed anything? Does `git_index_write` behave nicely?

File: LibGit2Sharp/Index.cs [hunk 2]
```diff
@@ -332,5 +349,18 @@ public virtual void Replace(Commit commit, IEnumerable<string> paths, ExplicitPa
             var changes = repo.Diff.Compare<TreeChanges>(commit.Tree, DiffTargets.Index, paths, explicitPathsOptions, new CompareOptions { Similarity = SimilarityOptions.None });
             Replace(changes);
         }
+
+        /// <summary>
+        /// Disposes the Index, provided it is not the Repository's index,
+        /// and was generated (for example, by merge).  If it is the
+        /// Repository's index, this has no effect.
+        /// </summary>
+        public void Dispose()
```
Comment (nulltoken) [inline LibGit2Sharp/Index.cs]: Maybe should we also add a finalizer?

File: LibGit2Sharp/Repository.cs (hunk 1 above)
Comment (ethomson) [inline LibGit2Sharp/Repository.cs]: Yeah, I think so as well.  Anything that hits `UpdatePhysicalIndex` is going to throw.  I think perhaps something that wraps an index handle but is not an index might be appropriate.  I think that you want to be able to:

* Get Conflicts
* Write it to a tree

...other things?

`ReadOnlyIndex`?

**Comment by ethomson** [comment]
@nulltoken did you have any thoughts on my question(s) about what to call this interface / concrete implementation and / or what additional things it should expose?

**Comment by nulltoken** [comment]
How about something like

```csharp
MergeTreeResult ObjectDatabase.MergeTree(Commit one, Commit another, MergeTreeOptions options)
```

Where `MergeTreeResult` would expose a status (Success, Conflicted) and either a resulting `Tree` or a list of conflicts?

**Comment by ethomson** [comment]
I have updated this with some changes.

**Comment by nulltoken** [comment]
Neat!

Could you please update the commit message of `Introduce ObjectDatabase.MergeCommits` so that it doesn't refer to the Index any longer?

File: LibGit2Sharp.Tests/MergeFixture.cs [hunk 3]
```diff
@@ -748,6 +748,98 @@ public void CanMergeIntoOrphanedBranch()
...
+        [Fact]
+        public void CanMergeTreeIntoSameTree()
+        {
+            string path = SandboxMergeTestRepo();
+            using (var repo = new Repository(path))
+            {
+                var master = repo.Branches["master"].Tip;
+
+                var result = repo.ObjectDatabase.MergeCommits(master, master, null);
+                Assert.Equal(MergeTreeStatus.Succeeded, result.Status);
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/MergeFixture.cs]: Assert.NotNull(MergeTreeStatus.Tree);
Assert.Equal(0, MergeTreeStatus.Conflicts.Count());

File: LibGit2Sharp.Tests/MergeFixture.cs [hunk 4]
```diff
@@ -748,6 +748,98 @@ public void CanMergeIntoOrphanedBranch()
...
+                repo.Index.Clear();
+                repo.Stage("README");
+
+                repo.Commit("A new world, free of the burden of the history", Constants.Signature, Constants.Signature);
+
+                var master = repo.Branches["master"].Tip;
+                var branch = repo.Branches["unborn"].Tip;
+
+                var result = repo.ObjectDatabase.MergeCommits(master, branch, null);
+                Assert.Equal(MergeTreeStatus.Succeeded, result.Status);
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/MergeFixture.cs]: Assert.NotNull(MergeTreeStatus.Tree);
Assert.Equal(0, MergeTreeStatus.Conflicts.Count());

File: LibGit2Sharp.Tests/MergeFixture.cs [hunk 5]
```diff
@@ -748,6 +748,98 @@ public void CanMergeIntoOrphanedBranch()
...
+
+                repo.Index.Replace(repo.Lookup<Commit>("conflicts"));
+
+                repo.Commit("A conflicting world, free of the burden of the history", Constants.Signature, Constants.Signature);
+
+                var master = repo.Branches["master"].Tip;
+                var branch = repo.Branches["unborn"].Tip;
+
+                var result = repo.ObjectDatabase.MergeCommits(master, branch, null);
+                Assert.Equal(MergeTreeStatus.Conflicts, result.Status);
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/MergeFixture.cs]: Assert.Null(MergeTreeStatus.Tree);
Assert.NotEqual(0, MergeTreeStatus.Conflicts.Count());

File: LibGit2Sharp.Tests/MergeFixture.cs [hunk 6]
```diff
@@ -748,6 +748,98 @@ public void CanMergeIntoOrphanedBranch()
...
+        public void CanMergeFastForwardTreeWithoutConflicts()
+        {
+            string path = SandboxMergeTestRepo();
+            using (var repo = new Repository(path))
+            {
+                var master = repo.Lookup<Commit>("master");
+                var branch = repo.Lookup<Commit>("fast_forward");
+
+                var result = repo.ObjectDatabase.MergeCommits(master, branch, null);
+                Assert.Equal(MergeTreeStatus.Succeeded, result.Status);
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/MergeFixture.cs]: Assert.NotNull(MergeTreeStatus.Tree);
Assert.Equal(0, MergeTreeStatus.Conflicts.Count());

File: LibGit2Sharp.Tests/MergeFixture.cs [hunk 7]
```diff
@@ -748,6 +748,98 @@ public void CanMergeIntoOrphanedBranch()
...
+        {
+            string path = SandboxMergeTestRepo();
+            using (var repo = new Repository(path))
+            {
+                var master = repo.Lookup<Commit>("master");
+                var branch = repo.Lookup<Commit>("conflicts");
+
+                var result = repo.ObjectDatabase.MergeCommits(master, branch, null);
+
+                Assert.Equal(MergeTreeStatus.Conflicts, result.Status);
```
Comment (nulltoken) [inline LibGit2Sharp.Tests/MergeFixture.cs]: Assert.Null(MergeTreeStatus.Tree);

**Comment by ethomson** [comment]
Updated with your suggestions, @nulltoken

**Comment by ethomson** [comment]
Oops, wrote that test sloppily.  Updated.

**Comment by nulltoken** [comment]
@ethomson Thanks for this. One final nitpick. https://github.com/libgit2/libgit2sharp/commit/b2b6139b1e3c45b00e53af75fb782586cae07665 commit message states

```
Introduce ObjectDatabase.MergeCommits

Provide an API to merge two commits and return the resultant result.
This Index may be analyzed looking for conflicts or otherwise to
determine the success of the merge.  This Index must be disposed when
the caller has finished with it.  Thus, Index now implements
IDisposable and Indexes that are not the repository index (ie, generated
by merge) are expected to be disposed.  The repository's index is
not expected to be disposed and is protected from disposal.
```

Could you please reword it so that it doesn't refer to the Index any longer?

**Comment by ethomson** [comment]
Yeah, I updated that sloppily.  Fixed.

**Comment by nulltoken** [comment]
Published as NuGet pre-release package `LibGit2Sharp.0.22.0-pre20150409142943`

**Pull request merged**